doc/source/templates/04_finetuning_llms_with_deepspeed/finetune_hf_llm.py

```diff
@@ -0,0 +1,569 @@
+        help="Learning rate to use."
+    )
+        
+    parser.add_argument(
+        "--as-test", 
+        action="store_true",
+        help="If passed, will run the script in test mode."
+    )
+
+    parser.add_argument(
+        "--ds-config", 
+        type=str,
+        default="./deepspeed_configs/zero_3_llama_2_7b.json", 
+        help="Deepspeed config json to use."
+    )
+    args = parser.parse_args()
+    
+    return args
+
+def main():
+    
+    args = parse_args()
+    
+    if not args.output_dir:
+        raise ValueError("--output_dir must be specified")
+    
+    # update the config with args so that we have access to them.
+    config = vars(args)
+    config.update(**{
+        "lr": args.lr, 
+        "num_epochs": args.num_epochs, 
+        "seed": 42, 
+        "batch_size": args.batch_size_per_device, 
+        "gradient_accumulation_steps": args.grad_accum,
+        "model_name": args.model_name,
+        "block_size": args.ctx_len,
+        "eval_batch_size": args.eval_batch_size_per_device,
+    })
+
+    # Add deepspeed plugin to the config
+    ds_plugin = DeepSpeedPlugin(hf_ds_config=config.get("ds_config"))
+    config.update(ds_plugin=ds_plugin)
+    
+    ray.init(
+        runtime_env={
+            "env_vars": {
+                "HF_HOME": "/mnt/local_storage/.cache/huggingface",
+                "TUNE_RESULT_DIR": args.output_dir,
+            }
+        }
+    )
+    
+    
+    train_ds = create_ray_dataset(args.train_path)
+    if args.test_path is not None:
+        valid_ds = create_ray_dataset(args.test_path)
+    else:
+        valid_ds = None
+    
+    # json file
+    with open(args.special_token_path, 'r') as json_file:
+        special_tokens = json.load(json_file)["tokens"]
+    
+    trainer = TorchTrainer(
+        training_function,
+        train_loop_config={
+            "config": config, 
+            "args": vars(args), 
+            "special_tokens": special_tokens
+        },
+        run_config=air.RunConfig(
+            # storage_path=args.output_dir,
+            # # For NFS: disables the default syncing to head node (already disabled on 2.6+)
+            # sync_config=tune.SyncConfig(syncer=None),
+            storage_path="s3://anyscale-staging-data-cld-kvedzwag2qa8i5bjxuevf5i7/templates-release-tests/04_finetuning_llms",
+            checkpoint_config=air.CheckpointConfig(
+                num_to_keep=1,
+                # Enable distributed checkpointing
+                _checkpoint_keep_all_ranks=True,
+                _checkpoint_upload_from_workers=True,
+            ),
+        ),
+        scaling_config=air.ScalingConfig(
+            num_workers=args.num_devices, 
+            use_gpu=True,  
+            resources_per_worker={"GPU": 1},
+        ),
+        datasets={
+            "train": train_ds,
+            "valid": valid_ds,
+        }
+    )
+
+    results = trainer.fit()
+    
+    print("results are stored in:")
+    print(results.checkpoint.uri)
```

Comment:
As part of this template, could you show what this directory ends up looking like? Would be helpful to mentally visualize what the final checkpoint looks like here!

Reply:
Adding this to readme.